Index: dnsmasq-2.62/src/dnsmasq.c
===================================================================
--- dnsmasq-2.62.orig/src/dnsmasq.c	2013-04-04 18:38:35.711255770 +0300
+++ dnsmasq-2.62/src/dnsmasq.c	2013-04-04 18:40:22.231258309 +0300
@@ -258,11 +258,12 @@
 	  script_uid = ent_pw->pw_uid;
 	  script_gid = ent_pw->pw_gid;
 	 }
-      else
-	baduser = daemon->scriptuser;
+//	else
+//	baduser = daemon->scriptuser;
     }
 #endif
-  
+
+#if 0  
   if (daemon->username && !(ent_pw = getpwnam(daemon->username)))
     baduser = daemon->username;
   else if (daemon->groupname && !(gp = getgrnam(daemon->groupname)))
@@ -270,6 +271,7 @@
 
   if (baduser)
     die(_("unknown user or group: %s"), baduser, EC_BADCONF);
+#endif
    
   /* implement group defaults, "dip" if available, or group associated with uid */
   if (!daemon->group_set && !gp)
