--- a/src/drivers/driver_wext.c
+++ b/src/drivers/driver_wext.c
@@ -1892,12 +1892,14 @@ static void wpa_driver_wext_disconnect(s
 		 * SIOCSIWMLME commands (or tries to associate automatically
 		 * after deauth/disassoc).
 		 */
+/*
 		for (i = 0; i < 32; i++)
 			ssid[i] = rand() & 0xFF;
 		if (wpa_driver_wext_set_ssid(drv, ssid, 32) < 0) {
 			wpa_printf(MSG_DEBUG, "WEXT: Failed to set bogus "
 				   "SSID to disconnect");
 		}
+*/
 	}
 }
 
