--- a/src/ap/hw_features.c
+++ b/src/ap/hw_features.c
@@ -106,19 +106,9 @@ int hostapd_get_hw_features(struct hosta
 			/*
 			 * Disable all channels that are marked not to allow
 			 * IBSS operation or active scanning.
-			 * Use radar channels only if the driver supports DFS.
 			 */
-			if ((feature->channels[j].flag &
-			     HOSTAPD_CHAN_RADAR) && dfs_enabled) {
-				dfs = 1;
-			} else if (feature->channels[j].flag &
-				   (HOSTAPD_CHAN_NO_IBSS |
-				    HOSTAPD_CHAN_PASSIVE_SCAN |
-				    HOSTAPD_CHAN_RADAR)) {
-				feature->channels[j].flag |=
-					HOSTAPD_CHAN_DISABLED;
-			}
-
+			if (feature->channels[j].flag & HOSTAPD_CHAN_NO_IBSS)
+				feature->channels[j].flag |= HOSTAPD_CHAN_DISABLED;
 			if (feature->channels[j].flag & HOSTAPD_CHAN_DISABLED)
 				continue;
 
